### Import required dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [2]:
# Create the path and import the data
full_student_data = os.path.join('../School_District_Analysis/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [3]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [4]:
# Check for null values
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [5]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [6]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [7]:
# Drop duplicated rows and verify removal
student_df = student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [8]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [9]:
# Examine the grade column to understand why it is not an int
student_df['grade']

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [10]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace('th', '')
student_df['grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [11]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [12]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [13]:
# Display the mean math score using the mean function
mean_math_score = student_df['math_score'].mean()
mean_math_score


64.67573326141189

In [14]:
# Store the minimum reading score as min_reading_score
min_reading_score = student_df["reading_score"].min()
min_reading_score

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [15]:
# Use loc to display the grade column
student_df['grade'].loc[1:]

1         9
2         9
3         9
5         9
6        12
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14830, dtype: int32

In [16]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[:3, [3, 4, 5]]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [17]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
student_df.loc[student_df["grade"] == 9].describe()


student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [18]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_score = student_df["reading_score"].min()
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [19]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School
student_df.loc[(student_df["school_name"] == "Dixon High School") & (student_df["grade"] == 10)][["school_name","reading_score"]]

school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[569 rows x 2 columns]

In [20]:
# Find the mean reading score for all students in grades 11 and 12 combined.

student_df.loc[student_df['grade'].isin([11,12])]['reading_score'].mean()

74.90038089192188

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [21]:
# Using the groupby and mean functions, find the average budget for each SCHOOL TYPE.
avg_budget_per_grade_by_school_type = student_df.groupby(['school_type']).mean()
avg_budget_per_grade_by_school_type.loc[:, ["school_budget"]]


school_budget
school_type               
Charter      872625.656236
Public       911195.558251

In [22]:
# Using the groupby and mean functions, find the average budget per GRADE for each SCHOOL TYPE.
avg_budget_per_grade_by_school_type = student_df.groupby(['school_type', 'grade']).mean()
avg_budget_per_grade_by_school_type.loc[:, ["school_budget"]]


school_budget
school_type grade               
Charter     9      863817.290130
            10     871823.608811
            11     874262.713649
            12     885096.335017
Public      9      926800.159528
            10     914715.360382
            11     900248.905136
            12     895952.915971

In [23]:
# Using the groupby and mean functions, look at the average reading and math scores per school type.
avg_by_school = student_df.groupby(['school_type']).mean()
avg_by_school.loc[:, ["reading_score", "math_score"]]

reading_score  math_score
school_type                           
Charter          72.450603   66.761883
Public           72.281219   62.951576

In [24]:
## Using the groupby and mean functions, look at the average reading and math scores per school type and grade.
avg_by_school = student_df.groupby(['school_type', 'grade']).mean()
avg_by_school.loc[:, ["reading_score", "math_score"]]

reading_score  math_score
school_type grade                           
Charter     9          68.676790   70.077874
            10         69.628822   66.443206
            11         80.596379   68.024735
            12         70.475421   60.212121
Public      9          69.687981   63.771066
            10         73.160461   63.764121
            11         73.486448   59.314337
            12         73.342380   63.568319

In [25]:
# Use the `groupby`, `count`, and `sort_values` functions to find the total number of students at each school and sort from most students to least students.

total_number_of_students_by_school = student_df.groupby(['school_name'])['student_id'].count().sort_values(ascending=False)

total_number_of_students_by_school

school_name
Montgomery High School    2038
Green High School         1961
Dixon High School         1583
Wagner High School        1541
Silva High School         1109
Woods High School         1052
Sullivan High School       971
Turner High School         846
Bowers High School         803
Fisher High School         798
Richard High School        551
Campos High School         541
Odonnell High School       459
Campbell High School       407
Chang High School          171
Name: student_id, dtype: int64

In [26]:
# Find the average math score by grade for each school type 

avg_by_school = student_df.groupby(['school_type', 'grade']).mean()
avg_by_school.loc[:, ["math_score"]]

math_score
school_type grade            
Charter     9       70.077874
            10      66.443206
            11      68.024735
            12      60.212121
Public      9       63.771066
            10      63.764121
            11      59.314337
            12      63.568319

In [27]:
#Compare district vs charter schools for budget, size, and scores.
avg_by_school = student_df.groupby(['school_type', 'grade', 'school_name', 'school_budget']).mean()

avg_by_school.loc[:, ["reading_score", "math_score"]]

reading_score  \
school_type grade school_name            school_budget                  
Charter     9     Campbell High School   960726             67.948333   
                  Chang High School      840132             62.395699   
                  Dixon High School      870334             70.743260   
                  Green High School      832670             69.449271   
                  Montgomery High School 893368             67.436729   
                  Richard High School    889699             65.220741   
            10    Campbell High School   960726             75.283333   
                  Chang High School      840132             76.042857   
                  Dixon High School      870334             67.781898   
                  Green High School      832670             73.962520   
                  Montgomery High School 893368             64.636111   
                  Richard High School    889699             65.857241   
            11    Campbell High School   960726             73.237500   
                  Chang High School      840132             78.683333   
                  Dixon High School      870334             84.822857   
                  Green High School      832670             84.828970   
                  Montgomery High School 893368             81.084146   
                  Richard High School    889699             62.953503   
            12    Campbell High School   960726             59.673016   
                  Chang High School      840132             63.825714   
                  Dixon High School      870334             83.836765   
                  Green High School      832670             76.310989   
                  Montgomery High School 893368             63.247798   
                  Richard High School    889699             80.813158   
Public      9     Bowers High School     848324             63.818750   
                  Campos High School     957299             62.584127   
                  Fisher High School     817615             86.050758   
                  Odonnell High School   959474             77.554167   
                  Silva High School      991918             67.492116   
                  Sullivan High School   961125             66.304845   
                  Turner High School     956438             78.798980   
                  Wagner High School     846745             76.117988   
                  Woods High School      912243             58.872014   
            10    Campos High School     957299             70.447030   
                  Fisher High School     817615             76.679758   
                  Odonnell High School   959474             72.503125   
                  Silva High School      991918             81.541826   
                  Sullivan High School   961125             85.070068   
                  Turner High School     956438             75.787234   
                  Wagner High School     846745             66.111849   
                  Woods High School      912243             63.248433   
            11    Bowers High School     848324             79.669867   
                  Campos High School     957299             74.885000   
                  Fisher High School     817615             62.041007   
                  Odonnell High School   959474             59.158182   
                  Silva High School      991918             76.156044   
                  Sullivan High School   961125             72.066055   
                  Turner High School     956438             68.236111   
                  Wagner High School     846745             78.662917   
                  Woods High School      912243             70.360000   
            12    Bowers High School     848324             76.597785   
                  Campos High School     957299             78.298507   
                  Fisher High School     817615             60.407143   
                  Odonnell High School   

In [28]:
# Using the groupby and mean functions, find the average budget per grade for each school type.
avg_budget_per_grade_by_school_type = student_df.groupby(by='school_type').mean()
avg_budget_per_grade_by_school_type.loc[:, ["school_budget"]]



school_budget
school_type               
Charter      872625.656236
Public       911195.558251

In [29]:
# Use groupby and mean to find the average reading and math scores for each school type.
avg_student_scores_by_grade = student_df.groupby(['school_type','grade']).mean()
avg_student_scores_by_grade.loc[:, ["math_score", "reading_score"]]


math_score  reading_score
school_type grade                           
Charter     9       70.077874      68.676790
            10      66.443206      69.628822
            11      68.024735      80.596379
            12      60.212121      70.475421
Public      9       63.771066      69.687981
            10      63.764121      73.160461
            11      59.314337      73.486448
            12      63.568319      73.342380

In [30]:
# Group by school, then grade before calculating the average scores
avg_by_school = student_df.groupby(['school_type', 'school_name', 'grade']).mean()
avg_by_school.loc[:, ["reading_score", "math_score"]]

reading_score  math_score
school_type school_name            grade                           
Charter     Campbell High School   9          67.948333   52.393333
                                   10         75.283333   73.711538
                                   11         73.237500   53.655469
                                   12         59.673016   60.260317
            Chang High School      9          62.395699   56.651613
                                   10         76.042857   78.571429
                                   11         78.683333   69.461111
                                   12         63.825714   67.237143
            Dixon High School      9          70.743260   69.932193
                                   10         67.781898   72.604569
                                   11         84.822857   77.299184
                                   12         83.836765   52.894485
            Green High School      9          69.449271   71.613557
                                   10         73.962520   67.116203
                                   11         84.828970   59.985340
                                   12         76.310989   58.891209
            Montgomery High School 9          67.436729   75.364343
                                   10         64.636111   53.211616
                                   11         81.084146   72.169970
                                   12         63.247798   64.661827
            Richard High School    9          65.220741   65.313333
                                   10         65.857241   67.160000
                                   11         62.953503   76.958599
                                   12         80.813158   52.615789
Public      Bowers High School     9          63.818750   48.724107
                                   11         79.669867   50.115467
                                   12         76.597785   58.418987
            Campos High School     9          62.584127   51.876984
                                   10         70.447030   62.311386
                                   11         74.885000   76.700000
                                   12         78.298507   73.410448
            Fisher High School     9          86.050758   77.369697
                                   10         76.679758   61.926586
                                   11         62.041007   54.622302
                                   12         60.407143   73.287245
            Odonnell High School   9          77.554167   60.579167
                                   10         72.503125   52.384375
                                   11         59.158182   66.054545
                                   12         68.960406   75.661929
            Silva High School      9          67.492116   75.406639
                                   10         81.541826   51.888747
                                   11         76.156044   66.927473
                                   12         59.425116   58.206047
            Sullivan High School   9          66.304845   60.679651
                                   10         85.070068   60.212925
                                   11         72.066055   65.071560
                                   12         81.170000   59.913333
            Turner High School     9          78.798980   64.135374
                                   10         75.787234   70.795508
                                   11         68.236111   52.587037
                                   12         75.147619   66.271429
            Wagner High School     9          76.117988   62.746341
                                   10         66.111849   69.222262
                                   11         78.662917   63.773333
                                   12         78.530769   61.693750
            Woods High School      9          58.872014   71.588737
                                   10         63.248433   67.657835
                              

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

During this project, we were tasked to analyze a dataset of information based on the characteristics and averages of 2 groups of schools : charter and public. The analysis considered the level of budget for each classification of schools but it would be good to conduct an analysis of the same grade across each school type and within each school type. For instance how does grade 12 from charter school compares with grade 12 in public school. How does grade 12 in one school compare with grade 12 in another school. Finally, is budget the main driver in the performance of schools? Why does it appear that students performed better in reading than math? 

Finally, an important fact that we could gather at the end of this analysis was that the average school budget, of public and charter schools, was 911195.558251 and 872625.656236, respectively, meaning in this case public schools have more budget than charter ones and this not standard, unless charter schools in this case are merely a subset of public schools but not real private schools with adequante funding.  

In [32]:
# Store the row with the maximum overall reading score as `max_reading_row`

max_reading_score = student_df["reading_score"].max()
max_reading_row = student_df.loc[student_df["reading_score"] == max_reading_score]
max_reading_row

student_id    student_name  grade         school_name  reading_score  \
3605     22589871    Jeremy Wells     11   Dixon High School          100.0   
5041     39568500  Robert Johnson     12   Dixon High School          100.0   
11813    63851186       Mary Rush     11   Silva High School          100.0   
17576    45424751      Robert Kim     12  Wagner High School          100.0   
17908    47461327   Morgan Vaughn     11  Wagner High School          100.0   

       math_score school_type  school_budget  
3605         91.7     Charter         870334  
5041         73.9     Charter         870334  
11813        63.8      Public         991918  
17576        72.8      Public         846745  
17908        61.4      Public         846745

In [33]:
# Use loc to look at all public schools; use describe on this subset
public_schools_df = student_df.loc[student_df["school_type"] == "Public"]
public_schools_df.describe()

student_id        grade  reading_score   math_score  school_budget
count  8.120000e+03  8120.000000    8120.000000  8120.000000    8120.000000
mean   6.951549e+07    10.362808      72.281219    62.951576  911195.558251
std    3.449048e+07     1.125523      15.449927    16.151433   60951.900575
min    1.001891e+07     9.000000      10.900000     5.300000  817615.000000
25%    3.957351e+07     9.000000      62.100000    52.500000  846745.000000
50%    6.920620e+07    10.000000      74.000000    63.350000  912243.000000
75%    9.896576e+07    11.000000      83.900000    74.400000  961125.000000
max    1.299966e+08    12.000000     100.000000   100.000000  991918.000000

In [34]:
# Use loc to look at all charter schools; use describe on this subset
public_schools_df = student_df.loc[student_df["school_type"] == "Charter"]
public_schools_df.describe()

student_id        grade  reading_score   math_score  school_budget
count  6.711000e+03  6711.000000    6711.000000  6711.000000    6711.000000
mean   7.004029e+07    10.346744      72.450603    66.761883  872625.656236
std    3.457614e+07     1.063143      14.948025    15.206059   33468.101539
min    1.000906e+07     9.000000      10.500000     3.700000  832670.000000
25%    4.006577e+07     9.000000      62.300000    56.900000  832670.000000
50%    7.011300e+07    10.000000      73.600000    67.800000  870334.000000
75%    9.961835e+07    11.000000      84.100000    77.500000  893368.000000
max    1.299997e+08    12.000000     100.000000   100.000000  960726.000000